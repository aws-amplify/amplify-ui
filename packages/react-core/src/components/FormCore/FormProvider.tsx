import React from 'react';
import {
  useForm as useReactHookForm,
  FormProvider as ReactHookFormProvider,
} from 'react-hook-form';

import { FormValues, FormHandle, FormProviderProps } from './types';

const DEFAULT_MODE = 'onTouched';

const FormProvider = React.forwardRef<
  FormHandle,
  FormProviderProps<FormValues>
>(function FormProvider<Values extends FormValues>(
  { children, defaultValues, mode = DEFAULT_MODE }: FormProviderProps<Values>,
  ref: React.ForwardedRef<FormHandle<Values>>
) {
  const formProviderProps = useReactHookForm<Values>({
    defaultValues,
    mode,
  });

  const { getValues, reset } = formProviderProps;
  React.useImperativeHandle(
    ref,
    () => ({ getValues, reset: () => reset(defaultValues) }),
    [defaultValues, getValues, reset]
  );

  return (
    <ReactHookFormProvider {...formProviderProps}>
      {children}
    </ReactHookFormProvider>
  );
});

export default FormProvider;
