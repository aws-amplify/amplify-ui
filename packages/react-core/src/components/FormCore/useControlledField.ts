import { useController } from 'react-hook-form';

import {
  FormValues,
  FocusHandler,
  UseControlledField,
  UseControlledFieldParams,
} from './types';
import { isTypedFunction } from '@aws-amplify/ui';

export const DEFAULT_ERROR_MESSAGE =
  '`useControlledField` must be used within a `FormProvider`';

/**
 * `Field` integration hook for usage with React Native `Field` components
 *
 * @param params Requires `name`, all additional params optional
 * @returns `Form` aware `Field` handlers and state values
 */
export default function useControlledField<OnBlur extends FocusHandler>({
  onBlur: _onBlur,
  onChangeText: _onChangeText,
  ...rest
}: UseControlledFieldParams<OnBlur>): UseControlledField {
  // If called outside a `FormProvider`, `useController` throws a `TypeError` instead
  // of failing gracefully (for example returning `undefined` or `null`).
  // Wrap call to `useController` in a `try`/`catch` for graceful failure
  let controller;
  try {
    controller = useController<FormValues>(rest);
  } catch {
    throw new Error(DEFAULT_ERROR_MESSAGE);
  }

  const {
    // `useController.onChange` handles `string` values passed directly to `onChange`
    // for RN support, map to `onChangeText`
    field: { name, onBlur, onChange: onChangeText, ref, value },
    fieldState: { error, isDirty, isTouched },
    formState: { isValid: invalid },
  } = controller;

  const errorMessage = error?.message;
  const hasError = !!errorMessage;

  const handleBlur = (event: Parameters<OnBlur>[0]) => {
    if (isTypedFunction(_onBlur)) {
      _onBlur(event);
    }
    // `useController.onBlur` does not receive params
    onBlur();
  };

  const handleChangeText = (event: string) => {
    if (isTypedFunction(_onChangeText)) {
      _onChangeText(event);
    }
    onChangeText(event);
  };

  return {
    errorMessage,
    hasError,
    invalid,
    isDirty,
    isTouched,
    name,
    onBlur: handleBlur,
    onChangeText: handleChangeText,
    ref,
    value,
  };
}
