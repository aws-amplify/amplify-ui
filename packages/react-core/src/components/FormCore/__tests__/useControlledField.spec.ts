import { act, renderHook } from '@testing-library/react-hooks';

import { useControlledField } from '..';
import FormProvider from '../FormProvider';
import { DEFAULT_ERROR_MESSAGE } from '../useControlledField';

describe('useControlledField', () => {
  it('returns the expected values in the happy path', async () => {
    const { result, waitForNextUpdate } = renderHook(
      () => useControlledField({ name: 'test-field' }),
      { wrapper: FormProvider }
    );

    await waitForNextUpdate();

    expect(result.current).toMatchSnapshot();
  });

  it('throws when called outside a FormProvider', () => {
    const { result } = renderHook(() =>
      useControlledField({ name: 'test-field' })
    );

    expect(result.error?.message).toBe(DEFAULT_ERROR_MESSAGE);
  });

  it('calls handlers passed as props as expected', async () => {
    const handleChangeText = jest.fn();
    const handleBlur = jest.fn();

    const { result, waitForNextUpdate } = renderHook(
      () =>
        useControlledField({
          name: 'test-field',
          onChangeText: handleChangeText,
          onBlur: handleBlur,
        }),
      { wrapper: FormProvider }
    );

    await waitForNextUpdate();

    const blurEvent = { text: 'blurrr' };

    act(() => {
      result.current.onBlur(blurEvent);
    });

    expect(handleBlur).toHaveBeenCalledTimes(1);
    expect(handleBlur).toHaveBeenCalledWith(blurEvent);

    const changeTextEvent = 'change';
    act(() => {
      result.current.onChangeText(changeTextEvent);
    });

    expect(handleChangeText).toHaveBeenCalledTimes(1);
    expect(handleChangeText).toHaveBeenCalledWith(changeTextEvent);
  });
});
