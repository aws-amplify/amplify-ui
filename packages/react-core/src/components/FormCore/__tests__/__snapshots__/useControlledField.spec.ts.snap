// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useControlledField returns the expected values in the happy path 1`] = `
{
  "errorMessage": undefined,
  "hasError": false,
  "invalid": true,
  "isDirty": false,
  "isTouched": false,
  "name": "test-field",
  "onBlur": [Function],
  "onChangeText": [Function],
  "ref": [Function],
  "value": undefined,
}
`;
